MyMovies/Sources/Features/MovieSearch/ViewModels/MovieSearchViewModel.swift

```diff
@@ -0,0 +1,120 @@
+/// - `failed`: The search operation failed, either due to a network error or no results.
+///
+/// ### Error Handling
+/// The `MovieSearchError` enum defines two possible error cases:
+/// - `networkError`: Indicates a network-related failure during the search.
+/// - `noResults`: Indicates that no movies were found for the given search query.
+///
+/// ### Example Usage
+/// ```swift
+/// let viewModel = MovieSearchViewModel(service: MovieService())
+/// viewModel.searchMovies(with: "Inception")
+/// print(viewModel.state) // Outputs the updated state after performing the search
+/// ```
+@Observable
+class MovieSearchViewModel<Service: Searchable>: ViewModelable {
+    
+    /// Errors that can occur during the movie search process.
+    enum MovieSearchError: Error, Equatable {
+        /// Indicates a network-related error.
+        case networkError
+        /// Indicates that no results were found for the given query.
+        case noResults
+    }
+    
+    /// The various states the view can be in during the search process.
+    enum State: Equatable {
+        /// The view is idle, with no ongoing search operation.
+        case idle
+        /// A search operation is currently in progress.
+        case loading
+        /// Movies have been successfully fetched.
+        case loaded(movies: [Movie])
+        /// The search operation failed with a specific error.
+        case failed(error: MovieSearchError)
+    }
+    
+    /// The current search text entered by the user.
+    var searchText: String = ""
+    
+    /// The current state of the search view.
+    var state: State = .idle
+    
+    /// The service used to perform the search operation.
+    private let service: Service
+    
+    /// Initializes the view model with the given service.
+    ///
+    /// - Parameter service: An instance of a type conforming to the `Searchable` protocol.
+    init(with service: Service) {
+        self.service = service
+    }
+    
+    /// A task used to handle the ongoing search operation.
+    private var searchTask: Task<Void, Never>?
+    
+    /// Initiates a search operation with the provided search text.
+    ///
+    /// - Parameter text: The search query entered by the user.
+    ///
+    /// The method:
+    /// - Cancels any ongoing search task.
+    /// - Sets the state to `.loading` while the search is performed.
+    /// - Updates the state to `.loaded` if movies are found, or `.failed` if no results or errors occur.
+    func searchMovies(with text: String) {
+        // Cancel any ongoing search task to avoid redundant or overlapping operations.
+        searchTask?.cancel()
+        
+        // Start a new search task.
+        searchTask = Task { [weak self] in
+            guard let self else { return }
+
+            do {
+                // Set the state to loading with animation for a better user experience.
+                withAnimation {
```

Comment:
Wrapping these state values in withAnimation is not something I love as it feels like specific "ux logic" but I dont think theres necessarily a better way to do this.